source/server/admin/stats_render.cc

```diff
@@ -105,31 +98,43 @@ void StatsTextRender::addDisjointBuckets(const std::string& name,
 
 StatsJsonRender::StatsJsonRender(Http::ResponseHeaderMap& response_headers,
                                  Buffer::Instance& response, const StatsParams& params)
-    : histogram_buckets_mode_(params.histogram_buckets_mode_) {
+    : histogram_buckets_mode_(params.histogram_buckets_mode_),
+      json_(std::make_unique<JsonContext>(response)), response_(response) {
   response_headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Json);
-  // We don't create a JSON data model for the entire stats output, as that
-  // makes streaming difficult. Instead we emit the preamble in the
-  // constructor here, and create json models for each stats entry.
-  response.add("{\"stats\":[");
-  histogram_array_ = std::make_unique<ProtobufWkt::ListValue>();
+}
+
+StatsJsonRender::JsonContext::JsonContext(Buffer::Instance& response)
+    : streamer_(response), stats_map_(streamer_.makeRootMap()) {
+  // We don't create a JSON data model for the stats output, as that makes
+  // streaming difficult. Instead we emit the preamble in the constructor here,
+  // and create json models for each stats entry.
+  stats_map_->addKey("stats");
+  stats_array_ = stats_map_->addArray();
+}
+
+void StatsJsonRender::drainIfNeeded(Buffer::Instance& response) {
+  if (&response_ != &response) {
+    response.move(response_);
+  }
 }
 
 // Buffers a JSON fragment for a numeric stats, flushing to the response
 // buffer once we exceed JsonStatsFlushCount stats.
 void StatsJsonRender::generate(Buffer::Instance& response, const std::string& name,
                                uint64_t value) {
-  response.addFragments({delim_, JsonNameTag, Json::sanitize(name_buffer_, name), JsonValueTag,
-                         std::to_string(value), JsonCloseBrace});
-  delim_ = ",";
+  ASSERT(!histograms_initialized_);
+  json_->stats_array_->addMap()->addEntries(
+      {{"name", name}, {"value", static_cast<double>(value)}});
```

Comment:
I think there could be loss of precision by converting to double. I think it would be better to add native int64/uint64 support to the json streamer.

Reply:
Great catch! If this is consumed by JavaScript this doesn't matter, I scanned JSON docs and it does not specify what the data type is in any language; just the syntax. So if we have another language reading json it could read high-precision integers and we should write them.

This also I think will be slightly faster.